docs: add instructions to build kindest-node image (#29079)
cilium-team/cilium: Documentation/contributing/development/dev_setup.rst

    $ make generate-k8s-api
    $ git add go.mod go.sum vendor/

Add/update a cilium/kindest-node image
--------------------------------------

Cilium might use its own fork of kindest-node so that it can use k8s versions

> that were not yet released by Kind maintainers.

Reviewer: ```suggestion
that have not been released by Kind maintainers yet.
```

> One other reason is the base image used on kindest-node, which has not been yet
> released: Debian Bookworm. Cilium requires Bookworm because the glibc version

Reviewer: We'll never think of updating this when Bookworm is out, let's make it an example rather than a specific reason?

```suggestion
One other reason for using a fork is that the base image used on kindest-node may not have been release yet. For example, as of this writing, Cilium requires Debian Bookworm (yet to be released), because the glibc version
```

> being available on Cilium's base docker image is the same as the one used in the
> Bookworm docker image which is relevant for testing with Golang's race detector.

Reviewer: ```suggestion
available on Cilium's base Docker image is the same as the one used in the
Bookworm Docker image which is relevant for testing with Go's race detector.
```

https://go.dev/doc/faq#go_or_golang

Currently, to publish an image on ``quay.io/cilium/kindest-node`` one has to be

> a maintainer. However, anyone can build a kindest-node image and try it out

Reviewer: ```suggestion
Currently, only maintainers can publish an image on ``quay.io/cilium/kindest-node``.
However, anyone can build a kindest-node image and try it out
```

on a Pull Request.

First, we need to build the base docker image and for that we need to clone the
kind repository https://github.com/kubernetes-sigs/kind and run

   .. code-block:: shell-session

>     make -C images/base/ quick

Reviewer: ```suggestion
To build a cilium/kindest-node image, first build the base Docker image:

   .. code-block:: shell-session

    git clone https://github.com/kubernetes-sigs/kind.git
    cd kind
    make -C images/base/ quick
```

> Note the resulting image tag for that command, it should be last tag built for

Reviewer: ```suggestion
Note the resulting image tag for that command, it should be the last tag built for
```

How do you check the tags for the repository? Would it make sense to turn it into a URL (https://gcr.io/k8s-staging-kind/base)?

Author: I've changed it to:
```
Take note of the resulting image tag for that command, it should be the last
tag built for the ``gcr.io/k8s-staging-kind/base`` repository in ``docker ps -a``.
```

the ``gcr.io/k8s-staging-kind/base`` repository.

Secondly, change into the directory with Kubernetes' source code which will be
used for the kindest node image. On this example, we will build a kindest-base

> image with Kubernetes version ``v1.28.3`` using the recently build base image

Reviewer: ```suggestion
image with Kubernetes version ``v1.28.3`` using the recently-built base image
```

``gcr.io/k8s-staging-kind/base:v20231108-a9fbf702``:

   .. code-block:: shell-session

    $ # Change to k8s' source code directory.

>     $ # git clone https://github.com/kubernetes/kubernetes.git
>     $ # cd kubernetes

Reviewer: Why the `#`? Users can just skip these if they've run it already, they're not comments?

```suggestion
    $ git clone https://github.com/kubernetes/kubernetes.git
    $ cd kubernetes
```

Author: fair point

    $ tag=v1.28.3
    $ git fetch origin --tags
    $ git checkout tags/${tag}
    $ kind build node-image \
      --image=quay.io/cilium/kindest-node:${tag} \
      --base-image=gcr.io/k8s-staging-kind/base:v20231108-a9fbf702

> Finally, publish the image on a public repository. If you are a maintainer and
> have permissions to publish on ``quay.io/cilium/kindest-node``, the renovate bot

Reviewer: ```suggestion
Finally, publish the image to a public repository. If you are a maintainer and
have permissions to publish on ``quay.io/cilium/kindest-node``, the Renovate bot
```
